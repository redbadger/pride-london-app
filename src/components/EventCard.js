// @flow
import React from "react";
import { View, StyleSheet, ImageBackground } from "react-native";
import formatDate from "date-fns/format";
import { lightNavyBlueColor } from "../constants/colors";
import Text from "./Text";
import SaveEventButton from "./SaveEventButton";
import { formattedEventPrice } from "../data/formatters";

type Props = {
  name: string,
  locationName: string,
  eventPriceLow: number,
  eventPriceHigh: number,
  startTime: string,
  endTime: string,
  imageUrl: string,
  isFree: boolean,
  isSaved: boolean,
  toggleSaved: boolean => void
};
const removeTimezoneFromDateString = isoString => isoString.slice(0, -6);

class EventCard extends React.PureComponent<Props> {
  static defaultProps = {
    isSaved: false
  };

  render() {
    const {
      name,
      locationName,
      startTime,
      endTime,
      imageUrl,
      eventPriceLow,
      eventPriceHigh,
      isFree,
      isSaved,
      toggleSaved
    } = this.props;
    const eventStartTime = removeTimezoneFromDateString(startTime);
    const eventEndTime = removeTimezoneFromDateString(endTime);
    const timeFormat = "HH:mm";
    const timeDisplay = `${formatDate(
      eventStartTime,
      timeFormat
    )} - ${formatDate(eventEndTime, timeFormat)}`;

    return (
      <View style={styles.eventCard}>
        <ImageBackground
          style={styles.imageContainer}
          source={{ uri: imageUrl }}
          resizeMode="cover"
        >
          <View style={styles.eventPriceContainer}>
            <Text type="price" color="whiteColor">
              {formattedEventPrice(isFree, eventPriceLow, eventPriceHigh)}
            </Text>
          </View>
        </ImageBackground>
        <View style={styles.eventCardDetails}>
          <View style={styles.column}>
            <View style={styles.row}>
              <View style={styles.column}>
                <Text
                  type="small"
                  color="lightNavyBlueColor"
                  style={styles.eventTime}
                >
                  {timeDisplay}
                </Text>
                <Text
                  numberOfLines={2}
                  type="h3"
                  color="lightNavyBlueColor"
                  style={styles.eventName}
                >
                  {name}
                </Text>
              </View>
              <View>
                <SaveEventButton active={isSaved} onPress={toggleSaved} />
              </View>
            </View>
            <Text type="small" color="lightNavyBlueColor">
              {locationName}
            </Text>
          </View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  row: {
    flex: 0,
    flexDirection: "row"
  },
  column: {
    flex: 0,
    flexDirection: "column",
    flexGrow: 1,
    flexShrink: 1
  },
  eventCard: {
    height: 108,
    flexDirection: "row",
    overflow: "hidden",
    borderRadius: 5
  },
  imageContainer: {
    width: 114,
    height: 108
  },
  eventPriceContainer: {
    height: 23,
    backgroundColor: lightNavyBlueColor,
    position: "absolute",
    paddingHorizontal: 5,
    justifyContent: "center"
  },
  eventCardDetails: {
    flex: 1,
    paddingLeft: 8
  },
  eventTime: {
    paddingTop: 12
  },
  eventName: {
    paddingTop: 4
  }
});

export default EventCard;
