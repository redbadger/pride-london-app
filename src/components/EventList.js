// @flow
import React from "react";
import {
  Animated,
  StyleSheet,
  SectionList,
  TouchableOpacity,
  View
} from "react-native";
import type { SectionBase } from "react-native/Libraries/Lists/SectionList";
import type { Props as ScrollViewProps } from "react-native/Libraries/Components/ScrollView/ScrollView";
import formatDate from "date-fns/format";
import EventCard from "./EventCard";
import Text from "./Text";
import type { Event, EventDays, Asset } from "../data/event";
import {
  eventListBgColor,
  eventListHeaderBgColor,
  eventListHeaderColor
} from "../constants/colors";

type Props = ScrollViewProps & {
  locale: string,
  events: EventDays,
  onPress: (eventName: string) => void,
  getAssetById: string => Asset
};

const separator = style => () => <View style={style} />;

type ItemProps = { item: Event };
const renderItem = (styles, locale, onPress, getAssetById) => ({
  item: event
}: ItemProps) => (
  <View style={styles.eventListItem}>
    <TouchableOpacity
      accessibilityTraits={["button"]}
      accessibilityComponentType="button"
      delayPressIn={50}
      onPress={() => onPress(event.sys.id)}
    >
      <EventCard
        name={event.fields.name[locale]}
        locationName={event.fields.locationName[locale]}
        price={event.fields.eventPriceLow[locale]}
        startTime={event.fields.startTime[locale]}
        endTime={event.fields.endTime[locale]}
        imageUrl={`https:${
          getAssetById(event.fields.eventsListPicture[locale].sys.id).fields
            .file[locale].url
        }`}
      />
    </TouchableOpacity>
  </View>
);

type Section = SectionBase<Event> & { title: string };
type SectionProps = { section: Section };
const renderSectionHeader = styles => ({ section }: SectionProps) => (
  <Text type="h2" style={styles.sectionHeader}>
    {section.title}
  </Text>
);

const eventSections = (events: EventDays, locale: string): Section[] =>
  events.map(it => ({
    data: it,
    title: formatDate(it[0].fields.startTime[locale], "dddd D MMMM")
  }));

const AnimatedSectionList = Animated.createAnimatedComponent(SectionList);
const EventList = ({
  contentContainerStyle,
  events,
  locale,
  onPress,
  getAssetById,
  ...otherScrollViewProps
}: Props) => (
  <AnimatedSectionList
    sections={eventSections(events, locale)}
    renderSectionHeader={renderSectionHeader(styles)}
    renderItem={renderItem(styles, locale, onPress, getAssetById)}
    keyExtractor={event => event.sys.id}
    contentContainerStyle={[styles.container, contentContainerStyle]}
    ItemSeparatorComponent={separator(styles.itemSeparator)}
    SectionSeparatorComponent={separator(styles.sectionSeparator)}
    stickySectionHeadersEnabled
    {...otherScrollViewProps}
  />
);

const styles = StyleSheet.create({
  itemSeparator: {
    height: 10
  },
  sectionSeparator: {
    height: 16
  },
  container: {
    paddingTop: 0,
    backgroundColor: eventListBgColor
  },
  eventListItem: {
    paddingHorizontal: 15
  },
  sectionHeader: {
    height: 32,
    paddingTop: 6,
    paddingBottom: 2,
    textAlign: "center",
    backgroundColor: eventListHeaderBgColor,
    color: eventListHeaderColor
  }
});

export default EventList;
