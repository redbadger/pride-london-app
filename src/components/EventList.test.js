import React from "react";
import { shallow } from "enzyme";
import EventList, { renderItem } from "./EventList";
import EventCard from "./EventCard";

const events = [
  [
    {
      sys: {
        id: "1",
        type: "",
        contentType: { sys: { id: "" } },
        revision: 1
      },
      fields: {
        name: {
          "en-GB": "some event"
        },
        startTime: {
          "en-GB": "2018-07-10T00:00"
        },
        endTime: {
          "en-GB": "2018-08-11T00:00"
        },
        locationName: {
          "en-GB": "Somewhere"
        },
        eventPriceLow: {
          "en-GB": "0"
        },
        eventPriceHigh: {
          "en-GB": "0"
        },
        imageUrl: {
          "en-GB": "http://placekitten.com/200/300"
        },
        isFree: {
          "en-GB": true
        }
      }
    },
    {
      sys: {
        id: "2",
        type: "",
        contentType: { sys: { id: "" } },
        revision: 1
      },
      fields: {
        name: {
          "en-GB": "some other"
        },
        startTime: {
          "en-GB": "2018-07-10T00:00"
        },
        endTime: {
          "en-GB": "2018-08-11T00:00"
        },
        locationName: {
          "en-GB": "Somewhere"
        },
        eventPriceLow: {
          "en-GB": "0"
        },
        eventPriceHigh: {
          "en-GB": "0"
        },
        imageUrl: {
          "en-GB": "http://placekitten.com/200/300"
        },
        isFree: {
          "en-GB": true
        }
      }
    }
  ],
  [
    {
      sys: {
        id: "2",
        type: "",
        contentType: { sys: { id: "" } },
        revision: 1
      },
      fields: {
        name: {
          "en-GB": "some other"
        },
        startTime: {
          "en-GB": "2018-07-11T00:00"
        },
        endTime: {
          "en-GB": "2018-08-11T00:00"
        },
        locationName: {
          "en-GB": "Somewhere"
        },
        eventPriceLow: {
          "en-GB": "0"
        },
        eventPriceHigh: {
          "en-GB": "0"
        },
        imageUrl: {
          "en-GB": "http://placekitten.com/200/300"
        },
        isFree: {
          "en-GB": true
        }
      }
    }
  ]
];

describe("EventList", () => {
  const render = props =>
    shallow(
      <EventList
        events={events}
        locale="en-GB"
        refreshing={false}
        onRefresh={() => {}}
        onPress={() => {}}
        getAssetById={() => {}}
        getAssetUrl={() => {}}
        savedEvents={new Set()}
        addSavedEvent={() => {}}
        removeSavedEvent={() => {}}
        {...props}
      />
    );

  it("renders correctly", () => {
    const output = render();
    expect(output).toMatchSnapshot();
  });

  it("renders section headers correctly", () => {
    const renderSectionHeader = render().prop("renderSectionHeader");
    const output = shallow(
      renderSectionHeader({ section: { title: "Hello" } })
    );

    expect(output).toMatchSnapshot();
  });

  describe("#renderItem", () => {
    it("renders items correctly", () => {
      const Item = renderItem({
        isSavedEvent: () => false,
        addSavedEvent: () => {},
        removeSavedEvent: () => {},
        locale: "en-GB",
        onPress: () => {},
        getAssetUrl: () => {}
      });
      const output = shallow(<Item item={events[0][0]} />);

      expect(output).toMatchSnapshot();
    });

    it("calls addSavedEvent when toggleSaved is called with true", () => {
      const event = events[0][0];
      const spy = jest.fn();
      const Item = renderItem({
        isSavedEvent: () => false,
        addSavedEvent: spy,
        removeSavedEvent: () => {},
        locale: "en-GB",
        onPress: () => {},
        getAssetUrl: () => {}
      });
      const output = shallow(<Item item={event} />);
      const toggleSaved = output.find(EventCard).prop("toggleSaved");
      toggleSaved(true);
      expect(spy).toBeCalledWith(event.sys.id);
    });

    it("calls removeSavedEvent when toggleSaved is called with false", () => {
      const event = events[0][0];
      const spy = jest.fn();
      const Item = renderItem({
        isSavedEvent: () => false,
        addSavedEvent: () => {},
        removeSavedEvent: spy,
        locale: "en-GB",
        onPress: () => {},
        getAssetUrl: () => {}
      });
      const output = shallow(<Item item={event} />);
      const toggleSaved = output.find(EventCard).prop("toggleSaved");
      toggleSaved(false);
      expect(spy).toBeCalledWith(event.sys.id);
    });
  });

  describe("#shouldComponentUpdate", () => {
    const props = {
      locale: "en-GB",
      refreshing: false,
      events,
      savedEvents: new Set()
    };

    it("stops update if locale, refresing and events stay the same", () => {
      const nextProps = {
        locale: "en-GB",
        refreshing: false,
        events: events.slice(0, 2), // force different instance
        savedEvents: props.savedEvents
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(false);
    });

    it("allows update when locale changes", () => {
      const nextProps = {
        locale: "en-US",
        refreshing: false,
        events,
        savedEvents: props.savedEvents
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(true);
    });

    it("allows update when refreshing", () => {
      const nextProps = {
        locale: "en-GB",
        refreshing: true,
        events,
        savedEvents: props.savedEvents
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(true);
    });

    it("allows update when events changue", () => {
      const nextProps = {
        locale: "en-GB",
        refreshing: false,
        events: events.slice(0, 1),
        savedEvents: props.savedEvents
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(true);
    });

    it("allows savedEvents change", () => {
      const nextProps = {
        locale: "en-GB",
        refreshing: false,
        events,
        savedEvents: new Set(["test"])
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(true);
    });
  });
});
