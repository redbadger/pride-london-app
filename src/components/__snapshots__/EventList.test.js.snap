// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<AnimatedComponent
  ItemSeparatorComponent={[Function]}
  SectionSeparatorComponent={[Function]}
  contentContainerStyle={
    Array [
      Object {
        "backgroundColor": "rgb(226, 226, 226)",
        "paddingTop": 0,
      },
      undefined,
    ]
  }
  keyExtractor={[Function]}
  onRefresh={[Function]}
  refreshing={false}
  renderItem={[Function]}
  renderSectionHeader={[Function]}
  sections={
    Array [
      Object {
        "data": Array [
          Object {
            "fields": Object {
              "name": Object {
                "en-GB": "some event",
              },
              "startTime": Object {
                "en-GB": "2018-07-10T00:00",
              },
            },
            "sys": Object {
              "contentType": Object {
                "sys": Object {
                  "id": "",
                },
              },
              "id": "1",
              "revision": 1,
              "type": "",
            },
          },
          Object {
            "fields": Object {
              "name": Object {
                "en-GB": "some other",
              },
              "startTime": Object {
                "en-GB": "2018-07-10T00:00",
              },
            },
            "sys": Object {
              "contentType": Object {
                "sys": Object {
                  "id": "",
                },
              },
              "id": "2",
              "revision": 1,
              "type": "",
            },
          },
        ],
        "title": "Tuesday 10 July",
      },
      Object {
        "data": Array [
          Object {
            "fields": Object {
              "name": Object {
                "en-GB": "some other",
              },
              "startTime": Object {
                "en-GB": "2018-07-11T00:00",
              },
            },
            "sys": Object {
              "contentType": Object {
                "sys": Object {
                  "id": "",
                },
              },
              "id": "2",
              "revision": 1,
              "type": "",
            },
          },
        ],
        "title": "Wednesday 11 July",
      },
    ]
  }
  stickySectionHeadersEnabled={true}
/>
`;
